timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/HistoryAnalyzer.java

```diff
@@ -0,0 +1,143 @@
+import com.palantir.timelock.history.PaxosAcceptorData;
+import com.palantir.timelock.history.models.CompletePaxosHistoryForNamespaceAndUseCase;
+import com.palantir.timelock.history.models.ConsolidatedLearnerAndAcceptorRecord;
+
+public final class HistoryAnalyzer {
+
+    public static boolean runCorruptionCheckOnHistory(CompletePaxosHistoryForNamespaceAndUseCase history) {
+        return verifyLearnersHaveLearnedSameValues(history)
+        && verifyLearnedValueWasAcceptedByQuorum(history)
+        && verifyLearnedValueIsGreatestAcceptedValue(history);
+    }
+
+    @VisibleForTesting
+    static boolean verifyLearnersHaveLearnedSameValues(CompletePaxosHistoryForNamespaceAndUseCase history) {
+        List<ConsolidatedLearnerAndAcceptorRecord> records = history.localAndRemoteLearnerAndAcceptorRecords();
+        return history.getAllSequenceNumbers()
+                .stream()
+                .allMatch(seq -> {
+                    Set<PaxosValue> learnedValuesForRound = getLearnedValuesForRound(records, seq);
+                    return learnedValuesForRound.size() <= 1;
+                });
+    }
+
+    @VisibleForTesting
+    static boolean verifyLearnedValueWasAcceptedByQuorum(CompletePaxosHistoryForNamespaceAndUseCase history) {
+        List<ConsolidatedLearnerAndAcceptorRecord> records = history.localAndRemoteLearnerAndAcceptorRecords();
+        int quorum = getQuorumSize(records);
+
+        return history.getAllSequenceNumbers()
+                .stream()
+                .allMatch(seq -> {
+                    Optional<PaxosValue> optionalLearnedValue = getLearnedValue(records, seq);
+                    if(!optionalLearnedValue.isPresent()) {
+                        return true;
+                    }
+
+                    PaxosValue learnedValue = optionalLearnedValue.get();
+                    List<PaxosValue> acceptedValues = getAcceptedValues(records, seq, learnedValue);
+                    return acceptedValues.size() >= quorum;
+                });
+    }
+
+    @VisibleForTesting
+    static boolean verifyLearnedValueIsGreatestAcceptedValue(CompletePaxosHistoryForNamespaceAndUseCase history) {
+        List<ConsolidatedLearnerAndAcceptorRecord> records = history.localAndRemoteLearnerAndAcceptorRecords();
+        return history.getAllSequenceNumbers()
+                .stream()
+                .allMatch(seq -> learnedValueIsGreatestAcceptedValue(records, seq));
+    }
+
+    private static boolean learnedValueIsGreatestAcceptedValue(
+            List<ConsolidatedLearnerAndAcceptorRecord> records, Long seq) {
+        byte[] learnedValueData = getPaxosValueData(getLearnedValue(records, seq));
+        if (learnedValueData == null) {
+            return true;
+        }
+
+        byte[] greatestAcceptedValueData = getPaxosValueData(getGreatestAcceptedValueAtSequence(records, seq));
+        if (greatestAcceptedValueData == null) {
+            // should not reach here
+            return false;
```

Comment:
if this is a bug in atlas code we should probably throw a SafeIllegalStateException